### PR TITLE
small code and docs refactoring

### PR DESCRIPTION
**Pull request by StrikerRUS**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -127,12 +127,6 @@ endif(USE_CUDA)
 if(USE_OPENMP)
     find_package(OpenMP REQUIRED)
     SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
-else()
-    # Ignore unknown #pragma warning
-    if((CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-      OR (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"))
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
```
Comment (StrikerRUS) [inline CMakeLists.txt:134]: Should be set after `-Wall` to take effect.

File: CMakeLists.txt [hunk 2]
```diff
@@ -272,6 +266,9 @@ if(UNIX OR MINGW OR CYGWIN)
     if(USE_SWIG)
         SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-strict-aliasing")
     endif()
+    if(NOT USE_OPENMP)
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas -Wno-unused-private-field")
```
Comment (StrikerRUS) [inline CMakeLists.txt:270]: Hide

```
/Users/runner/work/1/s/src/objective/binary_objective.hpp:208:14: warning: private field 'deterministic_' is not used [-Wunused-private-field]
  const bool deterministic_;
             ^
```

with AppleClang.

File: include/LightGBM/c_api.h [hunk 3]
```diff
@@ -408,6 +400,14 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetAddFeaturesFrom(DatasetHandle target,
 
 // --- start Booster interfaces
 
+/*!
+* \brief Get boolean representing whether booster is fitting linear trees.
+* \param handle Handle of booster
```
Comment (StrikerRUS) [inline include/LightGBM/c_api.h:405]: Should be `booster`, not `dataset`.

File: src/io/config.cpp [hunk 4]
```diff
@@ -340,9 +340,9 @@ void Config::CheckParamConflict() {
     Log::Warning("CUDA currently requires double precision calculations.");
     gpu_use_dp = true;
   }
-  // linear tree learner must be serial type and cpu device
+  // linear tree learner must be serial type and run on cpu device
   if (linear_tree) {
-    if (device_type == std::string("gpu")) {
+    if (device_type != std::string("cpu")) {
```
Comment (StrikerRUS) [inline src/io/config.cpp:345]: To generalize for possible future new enum options.

File: src/io/dataset_loader.cpp [hunk 5]
```diff
@@ -600,7 +600,6 @@ Dataset* DatasetLoader::LoadFromBinFile(const char* data_filename, const char* b
       }
       mem_ptr = buffer.data();
       const float* tmp_ptr_raw_row = reinterpret_cast<const float*>(mem_ptr);
-      std::vector<float> curr_row(dataset->num_numeric_features_, 0);
```
Comment (StrikerRUS) [inline src/io/dataset_loader.cpp:603]: Unused.

File: src/treelearner/linear_tree_learner.cpp [hunk 6]
```diff
@@ -211,7 +211,7 @@ void LinearTreeLearner::CalculateLinear(Tree* tree, bool is_refit, const score_t
     } else {
       raw_features = tree->branch_features(i);
     }
-    std::sort(raw_features.begin(), raw_features.end());
+    std::stable_sort(raw_features.begin(), raw_features.end());
```
Comment (StrikerRUS) [inline src/treelearner/linear_tree_learner.cpp:214]: Not sure about this. Refer to https://github.com/microsoft/LightGBM/pull/2154#issuecomment-490716927 and https://github.com/microsoft/LightGBM/pull/1739.
@btrotta Is there a possibility to meet equivalent elements in `raw_features`?
cc @guolinke

**Review by jameslamb — COMMENTED** [review]
changes look good to me! But @btrotta should approve before we merge

File: src/treelearner/linear_tree_learner.cpp (hunk 6 above)
Comment (guolinke) [reply src/treelearner/linear_tree_learner.cpp:214]: I think it is not need for stable-sort. I think here is only to get unique feature indices.

Comment (StrikerRUS) [reply src/treelearner/linear_tree_learner.cpp]: @guolinke 
Thanks! Reverted in 4358b4a783074fbd520e4882a79d1092ad9875ff.

**Review by guolinke — APPROVED** [review]

**Pull request merged**